### ISSUE TITLE
lang attribute of <html> tag ignores value of default locale and defaults to "en"

### ISSUE THREAD

**Issue by sephiroth-j** [vaadin/flow#13437]
### Description of the bug

The generated app shell contains `<html lang="en">` even if `Locale.getDefault()` returns "de_DE". This is different from Vaadin 14. No `I18NProvider` is used.

But even if one is used (such as [this example](https://github.com/vaadin/flow-spring-examples/blob/v23/src/main/java/org/vaadin/spring/tutorial/SimpleI18NProvider.java)), the value does not change.

### Expected behavior

The lang attribute of the <html> tag should follow the value of the default locale `Locale.getDefault()`.

### Minimal reproducible example

Run the application with `LANG="de_DE.utf8"`.

### Versions

- Vaadin / Flow version: 23.0.3
- Java version: 17.0.1
- OS version: Windows 10, Linux with LANG = de_DE

**Comment by sephiroth-j** [comment]
### Workaround (Spring Based)

Based on https://vaadin.com/docs/latest/flow/advanced/modifying-the-bootstrap-page/#IndexHtmlRequestListener-interface and https://vaadin.com/docs/latest/flow/advanced/service-init-listener  a possible workaround could look like this.

```java
@Bean
VaadinServiceInitListener vaadinServiceInitListener() {
	return event -> {
		event.addIndexHtmlRequestListener(indexHtmlResponse -> {
			// workaround for https://github.com/vaadin/flow/issues/13437
			indexHtmlResponse.getDocument().head().parent().attr("lang", Locale.getDefault().getLanguage());
		});
	};
}
```

**Comment by HermanBovens** [comment]
Note that the current "workaround" uses the system locale, which is usually not what you want in a multilingual app.  The actual language will need to come from the stored user settings, which you hopefully have access to in the service init listener.

**Comment by mshabarov** [comment]
The issue was triaged and currently added to the backlog priority queue for further investigation.

**Comment by czp13** [comment]
Thank you for reporting it, and for your patience. 

The relevant [fix](https://github.com/vaadin/flow/pull/16410) was developed, tested, added covering tests, and merged to the master, so the next Vaadin 24 release will contain it.

It is using the same logic as in the Component class for consistency within the framework:

0. It is checking if the language attribute is already added to the HTML tag, this was needed to not have any regression for customers having some solution in place.  If not, then it will try to add it based on the following rules:
1. trying to get the locale from the UI
2. if there is an I18N provider then the first locale will be used,
3. last option is to fallback to the Locale.getDefault()

Minor insight/comment: 
- Relying upon Locale.getDefault() is not a reliable way in my understanding as many factors, and variable can affect it, even the JDK version (Locale can have different implementations), for example in my JDK17 version I could not use any environmental variable to get used in the Locale only passed Java variables were manipulating it.
(so it was using System.getProperties but not System.getEnv() to init the Locale java core class values...)

for example, here we got the Java environment variables:
```
   private static Locale initDefault() {
   ... 
        language = props.getProperty("user.language", "en");
        // for compatibility, check for old user.region property
        region = props.getProperty("user.region");
```
More info in different `java.util.Locale implementations`. 
And I could not make it change with LC_ALL=en_US or LANG=en_US System variables. 
Maybe I missed something, but still I suggest not relying on Locale.getDefault().

**Comment by anm-cb** [comment]
@czp13 Thanks for tackling this issue, but is it really solved? I check with 23.3.11 and the language is always set from I18NProvider::getProvidedLocales because the UI is null in getLocale (there is a UI in the session-parameter of synchronizedHandleRequest() however).

**Comment by czp13** [comment]
@anm-cb Thanks for checking it. It is using the exact some logic as the Component class (so everything inherited from Component) for getting the Locale, it is unexpected that it is always null in your case, but still two more ways to manipulate the language (both for Component and for Lang attribute). You can get it from I18NProvider or for defaultLocale. Shall be sufficient for this case, as the reqoirement was to
give a way to manipulate lang attribute, not to give 3 different ways. It was just my solution to provide more options to do it and be consistent with other parts of the system.

Also I did a smoke test for one case and more tests in other versions, locale env, but I can have a deeper look next week and check why it is not getting it from UI for 23.3.7 flow.

Until then please use the other two ways if possible! thank you!(I am/will be on vacation until next week so limited way to reply)

**Comment by czp13** [comment]
Maybe I missed something with the backporting, merging, conflicts handling but as said I will have a look into the UI case of the all 3 cases.

**Comment by czp13** [comment]
@anm-cb 
Thank you for your patience! 

I spent a good 1-2 hours testing and investigating this matter in the past few days. I see, what is the problem you mentioning, and you were right, the UI is indeed created later in the flow (VaadinServlet, VaadinServletService, BootstrapHandler classes are the ones to look for).

If you feel this is a high priority problem, then please open an enhancement ticket/issue for fixing this issue in here (flow reopsitory) 🙏 🙇 

Note: 
Although after spending some hours:
-  I am not sure how to make sure that the UI is filled without a big refactor/change in logic order. 
-  I would not put this logic anywhere else, because it is related to the indexHtml other logic,
- Also, I would not change the logic from the component getLocal to have a seperate logic for index.html getLocal, as all changes in that static method will have a consistent change here and in other translations. Consistency is a good thing in my mind.

TLDR:
Until then there is still 2 ways to add a specific locale to the index.html which is an improvement I believe.

**Comment by mredeker** [comment]
We are using Vaadin 23.3.5 and had to use the workaround from [sephiroth-j](https://github.com/sephiroth-j)
Even when registering an I18N Provider the lang attribute of the html tag was always "en".

**Comment by knoobie** [comment]
Please take a look at the tags. It's fixed in 23.3.11.

**Comment by fxpaquette** [comment]
> We are using Vaadin 23.3.5 and had to use the workaround from [sephiroth-j](https://github.com/sephiroth-j) Even when registering an I18N Provider the lang attribute of the html tag was always "en".
> 
> This issue is marked as closed but it still seems to have an issue, or?

We had the same issue because the frontend/index.html we had was from an earlier version 23.xx) and still had the lang="en" tag in it. Delete your frontend/index.html file. It will be re-generated without the lang="en" tag and vaadin will use the I18Provider locale.
